yaml/shared/src/main/scala/org/virtuslab/yaml/YamlDecoder.scala

```diff
@@ -12,64 +12,96 @@ import org.virtuslab.yaml.Node.*
+    value.toLongOption.toRight(cannotParse(value, "Long", s))
   }
 
-  given YamlDecoder[Double] = YamlDecoder { case ScalarNode(value, _) =>
-    Try(value.toDouble).toEither
+  given YamlDecoder[Double] = YamlDecoder { case s @ ScalarNode(value, _) =>
+    value.toDoubleOption.toRight(cannotParse(value, "Double", s))
   }
 
-  given YamlDecoder[Float] = YamlDecoder { case ScalarNode(value, _) =>
-    Try(value.toFloat).toEither
+  given YamlDecoder[Float] = YamlDecoder { case s @ ScalarNode(value, _) =>
+    value.toFloatOption.toRight(cannotParse(value, "Float", s))
   }
 
-  given YamlDecoder[Short] = YamlDecoder { case ScalarNode(value, _) =>
-    Try(value.toShort).toEither
+  given YamlDecoder[Short] = YamlDecoder { case s @ ScalarNode(value, _) =>
+    value.toShortOption.toRight(cannotParse(value, "Short", s))
   }
 
-  given YamlDecoder[Byte] = YamlDecoder { case ScalarNode(value, _) =>
-    Try(value.toByte).toEither
+  given YamlDecoder[Byte] = YamlDecoder { case s @ ScalarNode(value, _) =>
+    value.toByteOption.toRight(cannotParse(value, "Byte", s))
   }
 
-  given YamlDecoder[Boolean] = YamlDecoder { case ScalarNode(value, _) =>
-    Try(value.toBoolean).toEither
+  given YamlDecoder[Boolean] = YamlDecoder { case s @ ScalarNode(value, _) =>
+    value.toBooleanOption.toRight(cannotParse(value, "Boolean", s))
+  }
+
+  given YamlDecoder[Any] = new YamlDecoder {
+    def construct(node: Node)(using settings: LoadSettings = LoadSettings.empty) = {
+      node match {
+        case ScalarNode(value, tag) if Tag.primitives.contains(tag) =>
+          tag match {
+            case Tag.nullTag => Right(None)
+            case Tag.boolean => value.toBooleanOption.toRight(cannotParse(value, "Boolean", node))
+            case Tag.int =>
+              if value.startsWith("0b") then
+                Try(Integer.parseInt(value.drop(2), 8)).toEither.swap
+                  .map(t => ConstructError.from(t, "Int", node))
+                  .swap
+              else if value.startsWith("0x") then
+                Try(Integer.parseInt(value.drop(2), 8)).toEither.swap
+                  .map(t => ConstructError.from(t, "Int", node))
+                  .swap
+              else value.toIntOption.toRight(cannotParse(value, "Int", node))
+            case Tag.float =>
+              value.toDoubleOption.toRight(cannotParse(value, "Double", node))
+            case Tag.str => Right(value)
+          }
+        case MappingNode(mappings, Tag.map) =>
+          val decoder = summon[YamlDecoder[Map[Any, Any]]]
+          decoder.construct(node)
+        case SequenceNode(seq, Tag.seq) =>
+          val decoder = summon[YamlDecoder[Seq[Any]]]
+          decoder.construct(node)
+        case _ =>
+          settings.constructors.get(node.tag) match {
+            case Some(decoder) => decoder.construct(node)
+            case None => Left(ConstructError(s"Could't create construct instance for ${node.tag}"))
```

Comment:
This message should contains more helpful text, for example that user should create own instance YamlDecoder for Tag.